File: C#/Unity/2018-2019/Unity WaveFunctionCollapse (Graduation)/Snippets/Entropy.cs
private Vector3Int MinEntropyCoords()
{
    float minEntropy = 0;
    Vector3Int minEntropyCoords = new Vector3Int();

    System.Random random = new System.Random();

    For3(outputSize, (x, y, z) =>
    {
        Vector3Int currentCoordinates = new Vector3Int(x, y, z);

        if (wave.MatrixData[x, y, z].AllowedCount() > 1)
        {
            float entropy = ShannonEntropy(currentCoordinates);
            float entropyPlusNoise = entropy - (float)random.NextDouble() / 1000;

            if (minEntropy == 0 || entropyPlusNoise < minEntropy)
            {
                minEntropy = entropyPlusNoise;
                minEntropyCoords = new Vector3Int(x, y, z);
            }
        }
    });

    return minEntropyCoords;
}

public float ShannonEntropy(Vector3Int currentCoordinates)
{
    int sumOfWeights = 0;
    float sumOfWeightsLogWeights = 0;

    foreach (var pair in wave.GetDataAt(currentCoordinates).allowedPatterns)
    {
        training.Weights.TryGetValue(pair.Key, out int weight);

        sumOfWeights += weight;
        sumOfWeightsLogWeights += weight * (float)Math.Log(weight);
    }
    return (float)Math.Log(sumOfWeights) - (sumOfWeightsLogWeights / sumOfWeights);
}

File: C++/2018-2019/WaveFunctionCollapse (Graduation)/Snippets/Entropy.cpp
FIntVector AWFC::MinEntropyCoords() {
	SCOPE_CYCLE_COUNTER( STAT_Entropy );

	float MinEntropy = 0;

	const FRandomStream RandomStream = FRandomStream( FMath::RandRange( 0.F, MAX_FLT ) );
	FIntVector RMinEntropyCoords = FIntVector();

	for3( OutputSize.X, OutputSize.Y, OutputSize.Z,
		  {
			  const FIntVector CurrentCoordinates = FIntVector( X, Y, Z );

			  if ( Wave.GetDataAt( CurrentCoordinates ).AllowedCount() > 1 ) {
				  const float Entropy = ShannonEntropy( CurrentCoordinates );
				  const float EntropyPlusNoise = Entropy - RandomStream.FRand();

				  if ( MinEntropy == 0 || EntropyPlusNoise < MinEntropy ) {
					  MinEntropy = EntropyPlusNoise;
					  RMinEntropyCoords = FIntVector( X, Y, Z );
				  }
			  }
		  }
	)

		return RMinEntropyCoords;
}

float AWFC::ShannonEntropy( FIntVector CurrentCoordinates ) {
	SCOPE_CYCLE_COUNTER( STAT_ShannonEntropy );

	int32 SumOfWeights = 0;
	int32 SumOfWeightsLogWeights = 0;

	for ( auto& Pair : Wave.GetDataAt( CurrentCoordinates ).AllowedPatterns ) {

		const int32 Weight = ModuleAssignee->Weights.FindRef( Pair.Key );

		SumOfWeights += Weight;
		SumOfWeightsLogWeights += Weight * static_cast< float >( FMath::Loge( Weight ) );
	}

	return FMath::Loge( SumOfWeights ) - ( SumOfWeightsLogWeights / SumOfWeights );
}
